From e9b6600041895b51e30443c9fd66ade99f0188d2 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Mon, 26 Feb 2024 13:22:38 +0100
Subject: [PATCH 0401/1695] drivers: stm32_rng: fix secure hardware RNG when
 using software RNG

Register STM32 RNG as non secure when CFG_WITH_SOFTWARE_PRNG is enabled
whatever CFG_PM value.
Register STM32 RNG to PM service when CFG_PM is enabled and
RNG is secure (that is CFG_WITH_SOFTWARE_PRNG is disabled).
Release STM32 RNG drier resource after core initialization when
it is not needed neither at runtime (is non-secure) nor during
STM32MP15 PM sequence.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 21877b9ecc68ec04500d410a752d4d9da95f6d49 drivers: stm32_rng: fix secure hardware RNG when using software RNG]
Change-Id: I768f3baf03e911a8330f33ede1ab5afe1461e6ef
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/365923
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 core/drivers/stm32_rng.c | 14 ++++++--------
 1 file changed, 6 insertions(+), 8 deletions(-)

diff --git a/core/drivers/stm32_rng.c b/core/drivers/stm32_rng.c
index 8e15a1930..622094087 100644
--- a/core/drivers/stm32_rng.c
+++ b/core/drivers/stm32_rng.c
@@ -93,7 +93,6 @@ struct stm32_rng_instance {
 	uint32_t health_test_conf;
 	uint32_t noise_ctrl_conf;
 	uint32_t rng_config;
-	bool release_post_boot;
 	bool clock_error;
 	bool error_conceal;
 };
@@ -597,10 +596,6 @@ static TEE_Result stm32_rng_parse_fdt(const void *fdt, int node)
 	if (fdt_getprop(fdt, node, "clock-error-detect", NULL))
 		stm32_rng->clock_error = true;
 
-	/* Release device if not used at runtime or for pm transitions */
-	stm32_rng->release_post_boot = IS_ENABLED(CFG_WITH_SOFTWARE_PRNG) &&
-				       !IS_ENABLED(CFG_PM);
-
 	stm32_rng->rng_config = stm32_rng->ddata->cr;
 	if (stm32_rng->rng_config & ~RNG_CR_ENTROPY_SRC_MASK)
 		panic("Incorrect entropy source configuration");
@@ -669,7 +664,7 @@ static TEE_Result stm32_rng_probe(const void *fdt, int offs,
 	if (stm32_rng->bus_clock)
 		clk_disable(stm32_rng->bus_clock);
 
-	if (stm32_rng->release_post_boot)
+	if (IS_ENABLED(CFG_WITH_SOFTWARE_PRNG))
 		stm32mp_register_non_secure_periph_iomem(stm32_rng->base.pa);
 	else
 		stm32mp_register_secure_periph_iomem(stm32_rng->base.pa);
@@ -678,7 +673,9 @@ static TEE_Result stm32_rng_probe(const void *fdt, int offs,
 	assert(stm32_rng->ddata->has_power_optim ==
 	       stm32_rng->ddata->has_cond_reset);
 
-	register_pm_core_service_cb(stm32_rng_pm, &stm32_rng, "rng-service");
+	if (!IS_ENABLED(CFG_WITH_SOFTWARE_PRNG))
+		register_pm_core_service_cb(stm32_rng_pm, &stm32_rng,
+					    "rng-service");
 
 	return TEE_SUCCESS;
 
@@ -742,7 +739,8 @@ DEFINE_DT_DRIVER(stm32_rng_dt_driver) = {
 
 static TEE_Result stm32_rng_release(void)
 {
-	if (stm32_rng && stm32_rng->release_post_boot) {
+	if (stm32_rng && IS_ENABLED(CFG_WITH_SOFTWARE_PRNG) &&
+	    !IS_ENABLED(CFG_STM32MP1_OPTEE_IN_SYSRAM)) {
 		DMSG("Release RNG driver");
 		free(stm32_rng);
 		stm32_rng = NULL;
-- 
2.47.2

